### PR TITLE
GH-10232 improved timezone selector

### PR DESCRIPTION
**Pull request by IndushaS**
Summary
Improved timezone selector located in account settings -> display -> timezone

1. Updated manual timezone component to a drop down list:
2. When "Automatic" is selected, dropdown list is disabled and automatic value is displayed. 

GitHub issue:
Fixes https://github.com/mattermost/mattermost-server/issues/10232

Screenshots:
image
image
image

### CODE REVIEW

**Review by esethna — APPROVED** [review]
Thanks @IndushaS, is there a follow-up ticket for updating the timezone list?

**Comment by esethna** [comment]
@IndushaS please note there should be some padding between the checkbox and the dropdown list:

image

**Comment by IndushaS** [comment]
Hello Eric,

To my knowledge there is no ticket for that yet. Also thank you for taking a look at the PR, I must have not seen the padding I'll be sure to fix that sorry.

Kindest Regards,
Indusha
________________________________

**Comment by esethna** [comment]
Thanks, please share the follow-up ticket once opened :)

**Comment by IndushaS** [comment]
> Thanks, please share the follow-up ticket once opened :)

Hello @esethna, I added the padding as you mentioned. I'm not sure how to create a ticket maybe you could tell me the general steps?

**Comment by gabrieljackson** [comment]
@IndushaS and @esethna, I will create the follow-up ticket and let you know when it is ready.

**Comment by emilyacook** [comment]
@gabrieljackson gentle ping :)

**Comment by gabrieljackson** [comment]
@emilyacook Sorry about the delay. New issue created here: https://github.com/mattermost/mattermost-server/issues/16918

File: components/user_settings/display/manage_timezones/index.ts [hunk 1]
```diff
@@ -18,6 +20,11 @@ function mapDispatchToProps(dispatch: Dispatch) {
             updateMe,
         }, dispatch)};
 }
+function mapStateToProps(state:GlobalState) {
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/index.ts:23]: nit: spaces

File: components/user_settings/display/manage_timezones/index.ts [hunk 2]
```diff
@@ -18,6 +20,11 @@ function mapDispatchToProps(dispatch: Dispatch) {
...
+    return {
+       times:getSupportedTimezones(state),
+    };
+}
 
-export default connect(null, mapDispatchToProps)(ManageTimezones);
+export default connect(mapStateToProps,mapDispatchToProps)(ManageTimezones);
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/index.ts:29]: Same here. Can you configure your eslint to match our configuration?

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 3]
```diff
@@ -26,7 +26,13 @@ type Props ={
     manualTimezone: string;
     timezones: string[];
     actions: Actions;
+    times:string[];
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:29]: nit: space

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 4]
```diff
@@ -35,24 +41,35 @@ type State ={
...
+        
 
         this.state = {
             useAutomaticTimezone: props.useAutomaticTimezone,
             automaticTimezone: props.automaticTimezone,
             manualTimezone: props.manualTimezone,
             manualTimezoneInput: props.manualTimezone,
             isSaving: false,
+            openMenu: false,
+            selectedOption: {label:props.manualTimezone, value: props.manualTimezone},
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:60]: nit: space

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 5]
```diff
@@ -35,24 +41,35 @@ type State ={
...
         };
     }
 
-    onChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
-        this.setState({manualTimezoneInput: e.target.value});
+    onChange = (selectedOption:ValueType<SelectedOption>) => {
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:64]: nit: spaces

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 6]
```diff
@@ -147,9 +164,48 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
...
     render() {
+        const timeOptions = this.props.times.map((timeString) => {
+            return {
+                value: timeString,
+                label: timeString
+              }
+            }
+          );
         const {timezones} = this.props;
+        console.log(this.props.times);
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:208]: nit: remove console.log

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 7]
```diff
@@ -147,9 +164,48 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
     handleManualTimezone = (e: React.ChangeEvent<HTMLSelectElement>) => {
         this.setState({manualTimezone: e.target.value});
     };
+    handleContainerKeyDown = (e: KeyboardEvent) => {
+        const modalBody = document.querySelector('.modal-body');
+        if (isKeyPressed(e, Constants.KeyCodes.ESCAPE) && this.state.openMenu) {
+            modalBody?.classList.remove('no-scroll');
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:170]: Please avoid using query selector and use classNames, We’re aiming to have a declarative codebase where possible. Also we have a pattern of identifying model you can take a look at examples in `edit_channel_header_modal.jsx`

We have close and open actions for modals as well as modal identifiers here 
`import {openModal, closeModal} from 'actions/views/modals';`

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 8]
```diff
@@ -147,9 +164,48 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
...
+            this.setState({openMenu: false});
+            e.stopPropagation();
+        }
+    };
+
+    handleKeyDown = (e: React.KeyboardEvent) => {
+        const modalBody = document.querySelector('.modal-body');
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:177]: Please avoid using query selector and use classNames, We’re aiming to have a declarative codebase where possible. Also we have a pattern of identifying model you can take a look at examples in edit_channel_header_modal.jsx

We have close and open actions for modals as well as modal identifiers here
import {openModal, closeModal} from 'actions/views/modals';

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 9]
```diff
@@ -147,9 +164,48 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
...
+        if (isKeyPressed(e, Constants.KeyCodes.ENTER)) {
+            modalBody?.classList.add('no-scroll');
+            this.setState({openMenu: true});
+        }
+    };
+    handleMenuClose = () => {
+        const modalBody = document.querySelector('.modal-body');
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:184]: same here

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 10]
```diff
@@ -147,9 +164,48 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
...
+        if (modalBody) {
+            modalBody.classList.remove('no-scroll');
+        }
+        this.setState({openMenu: false});
+    };
+
+    handleMenuOpen = () => {
+        const modalBody = document.querySelector('.modal-body');
```
Comment (nevyangelova) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:192]: and here

**Review by nevyangelova — CHANGES_REQUESTED** [review]

**Comment by devinbinnie** [comment]
Hey @IndushaS, thanks for your contribution! Couple things I'd like you to do before I review your PR:

1) Please sign the Contributor License Agreement found here: http://www.mattermost.org/mattermost-contributor-agreement/. When that's done, comment `/check-cla`.
2) Run `make check-style` and `make i18n-extract` and make sure there are no lint issues or missing translations. It looks like there are some errors around those.
3) Merge the latest `master` into your branch and fix any merge conflicts.

File: components/user_settings/display/manage_timezones/manage_timezones.tsx [hunk 11]
```diff
@@ -162,13 +173,21 @@ export default class ManageTimezones extends React.PureComponent<Props, State> {
 
         const inputs = [];
 
-        const timezoneRegion = (
+        /*const timezoneRegion = (
```
Comment (devinbinnie) [inline components/user_settings/display/manage_timezones/manage_timezones.tsx:176]: We can probably just remove this commented code

**Review by devinbinnie — APPROVED** [review]
LGTM! Thanks @IndushaS!
Just one non-blocking comment

**Review by nevyangelova — APPROVED** [review]
Thank you 🎉 looks great

**Review by jgilliam17 — APPROVED** [review]
Thank you @IndushaS 
Tested, looks good to merge.
- Verified time zone list drop down; when "Automatic" is selected, dropdown is disabled and shows the auto selected timezone in it. 
The timezone list will be updated/addressed in the follow up issue MM-32804.

**Pull request merged**